File: machine/chatbot/python/test_model.py
import numpy as np
import random
from tensorflow.keras.models import load_model
import json
from de_contract import *


def bag_of_words(message):
    with open("words.json", encoding='utf-8') as file:
        words = json.load(file)
    message = replace(message)
    message_words = message.split()
    bag = [0] * len(words)

    for word in message_words:
        for i, w in enumerate(words):
            if word == w:
                bag[i] = 1

    return np.array(bag)


def predict_class(message):
    model = load_model("model.h5")
    with open("classes.json", encoding='utf-8') as file:
        classes = json.load(file)
    
    response = model.predict(np.array([bag_of_words(message)]))[0]
    results = [[i, r] for i, r in enumerate(response)]
    results.sort(key=lambda x: x[1], reverse=True)

    return classes[results[0][0]]


def get_response(message, dataset):
    tag = predict_class(message)
    for data in dataset:
        if data['tag'] == tag:
            return random.choice(data['responses'])

File: machine/chatbot/javascript/test_model.js
const tf = require("@tensorflow/tfjs-node-gpu");
const { replace } = require("./de_contract");
const dataset = require("./dataset");
const words = require("./words.json");
const classes = require("./classes.json");

const bag_of_words = (message) => {
  message = replace(message);
  message = message.split(" ");
  let bag = Array(words.length).fill(0);
  for (word of message) {
    if (words.includes(word)) {
      bag[words.indexOf(word)] = 1;
    }
  }
  return bag;
};

const predict_class = async (message) => {
  const model = await tf.loadLayersModel(
    "file://machine/chatbot/javascript/model.json"
  );
  const predict = model.predict(tf.tensor([bag_of_words(message)]));
  const result = await predict.argMax(1).dataSync()[0];
  return classes[result];
};

const get_response = async (message) => {
  const tag = await predict_class(message);
  for (data of dataset) {
    if (data.tag === tag) {
      return data.responses[Math.floor(Math.random() * data.responses.length)];
    }
  }
};

module.exports = get_response;
